File: src/reflection_service/extraction.py
import asyncio
import logging
from dataclasses import replace
from uuid import UUID

from reflection_service.clients import EmbeddingClient, ModelClient
from reflection_service.db import ClaimedJob, Database
from reflection_service.domain import (
    EntityCandidate,
    MentionContext,
    PreparedClaim,
    PreparedEntity,
    PreparedSegment,
    claim_id_for,
    equivalence_key,
    new_entity_id_for,
    normalize_name,
    validate_claim_decisions,
    validate_resolutions,
)
from reflection_service.models import ExtractedClaim, Resolution

logger = logging.getLogger(__name__)


class ExtractionEngine:
    def __init__(
        self, database: Database, models: ModelClient, embeddings: EmbeddingClient
    ) -> None:
        self._database = database
        self._models = models
        self._embeddings = embeddings

    async def prepare(self, job: ClaimedJob) -> PreparedSegment:
        prior_summaries = await self._database.prior_summaries(
            job.request.session_id, job.segment_id
        )
        extracted = await self._models.extract(job.request, prior_summaries)

        context_specs: list[tuple[str, str, str, str]] = []
        for index, claim in enumerate(extracted.claims):
            supporting_claim = self._claim_text(claim)
            context_specs.append((f"c{index}.subject", "subject", claim.subject, supporting_claim))
            if claim.object_entity is not None:
                context_specs.append(
                    (f"c{index}.object", "object", claim.object_entity, supporting_claim)
                )

        candidate_inputs = [
            (
                f"Entity mention: {text}\nEndpoint role: {role}\n"
                f"Supporting claim: {supporting_claim}\nSegment summary: {extracted.summary}"
            )
            for _, role, text, supporting_claim in context_specs
        ]
        mention_embeddings = await self._embeddings.embed(candidate_inputs, input_type="query")
        candidate_sets = await asyncio.gather(
            *(
                self._database.entity_candidates(text, embedding)
                for (_, _, text, _), embedding in zip(
                    context_specs, mention_embeddings, strict=True
                )
            )
        )
        contexts = tuple(
            MentionContext(
                mention_id=mention_id,
                role=role,
                text=text,
                supporting_claim=supporting_claim,
                candidates=candidates,
            )
            for (mention_id, role, text, supporting_claim), candidates in zip(
                context_specs, candidate_sets, strict=True
            )
        )
        if contexts:
            resolution_result = await self._models.resolve(
                extracted.summary, extracted.claims, contexts
            )
            triaged_claims = validate_claim_decisions(extracted.claims, resolution_result)
            kept_context_ids = {f"c{index}.subject" for index, _ in triaged_claims}
            kept_context_ids.update(
                f"c{index}.object"
                for index, claim in triaged_claims
                if claim.object_entity is not None
            )
            kept_contexts = tuple(
                context for context in contexts if context.mention_id in kept_context_ids
            )
            validated = validate_resolutions(kept_contexts, resolution_result)
            logger.info(
                "claim triage completed proposed=%d kept=%d dropped=%d",
                len(extracted.claims),
                len(triaged_claims),
                len(extracted.claims) - len(triaged_claims),
            )
        else:
            triaged_claims = []
            kept_contexts = ()
            validated = {}

        entities_by_id: dict[UUID, PreparedEntity] = {}
        occurrence_entities: dict[str, PreparedEntity] = {}
        for context in kept_contexts:
            _, resolution = validated[context.mention_id]
            candidate = self._selected_candidate(context, resolution)
            if candidate is None:
                entity_id = new_entity_id_for(job.segment_id, resolution.canonical_name)
                canonical_name = resolution.canonical_name
                description = resolution.description
                is_new = True
            else:
                entity_id = candidate.id
                canonical_name = candidate.canonical_name
                description = candidate.description
                is_new = False
            aliases = self._unique_aliases(
                canonical_name,
                context.text,
                *resolution.aliases,
                *(candidate.aliases if candidate else ()),
            )
            previous = entities_by_id.get(entity_id)
            if previous is not None:
                entity = replace(
                    previous,
                    aliases=self._unique_aliases(*previous.aliases, *aliases),
                )
            else:
                entity = PreparedEntity(
                    id=entity_id,
                    canonical_name=canonical_name,
                    normalized_name=normalize_name(canonical_name),
                    description=description,
                    aliases=aliases,
                    embedding=None,
                    is_new=is_new,
                )
            entities_by_id[entity_id] = entity
            occurrence_entities[context.mention_id] = entity

        entities = list(entities_by_id.values())
        new_entities = [entity for entity in entities if entity.is_new]
        claim_texts = [self._claim_text(claim) for _, claim in triaged_claims]
        entity_texts = [f"{entity.canonical_name}: {entity.description}" for entity in new_entities]
        document_embeddings = await self._embeddings.embed(
            [*claim_texts, *entity_texts], input_type="document"
        )
        claim_embeddings = document_embeddings[: len(triaged_claims)]
        entity_embeddings = document_embeddings[len(triaged_claims) :]
        embedding_by_entity = {
            entity.id: embedding
            for entity, embedding in zip(new_entities, entity_embeddings, strict=True)
        }
        entities = [
            replace(entity, embedding=embedding_by_entity.get(entity.id)) for entity in entities
        ]

        claims: list[PreparedClaim] = []
        for (index, claim), embedding in zip(triaged_claims, claim_embeddings, strict=True):
            subject_entity = occurrence_entities[f"c{index}.subject"]
            object_entity = (
                occurrence_entities[f"c{index}.object"] if claim.object_entity is not None else None
            )
            claims.append(
                PreparedClaim(
                    id=claim_id_for(job.segment_id, index),
                    subject=claim.subject,
                    subject_entity_id=subject_entity.id,
                    predicate=claim.predicate,
                    confidence=claim.confidence,
                    object_entity=claim.object_entity,
                    object_entity_id=object_entity.id if object_entity is not None else None,
                    object_value=claim.object_value,
                    equivalence_key=equivalence_key(
                        subject_entity.id,
                        claim.predicate,
                        object_entity_id=(object_entity.id if object_entity is not None else None),
                        object_value=claim.object_value,
                    ),
                    embedding=embedding,
                )
            )
        return PreparedSegment(
            id=job.segment_id,
            session_id=job.request.session_id,
            start_user_message_id=job.request.start_user_message_id,
            end_user_message_id=job.request.end_user_message_id,
            summary=extracted.summary,
            entities=tuple(entities),
            claims=tuple(claims),
            projection_version=job.request.projection_version,
        )

    @staticmethod
    def _selected_candidate(
        context: MentionContext, resolution: Resolution
    ) -> EntityCandidate | None:
        if resolution.candidate_entity_id is None:
            return None
        return next(
            candidate
            for candidate in context.candidates
            if candidate.id == resolution.candidate_entity_id
        )

    @staticmethod
    def _claim_text(claim: ExtractedClaim) -> str:
        object_text = claim.object_entity or claim.object_value
        return f"{claim.subject} | {claim.predicate} | {object_text}"

    @staticmethod
    def _unique_aliases(*aliases: str) -> tuple[str, ...]:
        unique: dict[str, str] = {}
        for alias in aliases:
            unique.setdefault(normalize_name(alias), alias.strip())
        return tuple(unique.values())

File: server/src/extraction.ts
import { type ExtractedClaim } from "@reflection/shared/contracts";
import {
  claimIdFor,
  equivalenceKey,
  newMentionEntityIdFor,
  normalizeName,
  type MentionContext,
  type PreparedClaim,
  type PreparedEntity,
  type PreparedSegment,
  type ValidatedResolutionPlan,
} from "@reflection/shared/domain";

import {
  MAX_EMBEDDING_INPUT_BYTES,
  EmbeddingClient,
  ModelClient,
} from "./clients.js";
import type { ClaimedJob, Database } from "./database.js";
import type { ValidatedExtractionResult } from "./extraction-validation.js";

const MAX_ENTITY_DESCRIPTION_CODE_POINTS = 2_000;

function truncateCodePoints(value: string, maximum: number): string {
  if ([...value].length <= maximum) return value;
  return [...value].slice(0, maximum).join("");
}

function truncateUtf8Bytes(value: string, maximum: number): string {
  if (Buffer.byteLength(value, "utf8") <= maximum) return value;
  const retained: string[] = [];
  let bytes = 0;
  for (const character of value) {
    const characterBytes = Buffer.byteLength(character, "utf8");
    if (bytes + characterBytes > maximum) break;
    retained.push(character);
    bytes += characterBytes;
  }
  return retained.join("");
}

type ExtractionDatabase = Pick<Database, "priorSummaries" | "entityCandidates">;
type ExtractionModels = Pick<ModelClient, "extract" | "resolve">;
type ExtractionEmbeddings = Pick<EmbeddingClient, "embed">;

interface ContextSpec {
  mentionId: string;
  role: string;
  text: string;
  supportingClaim: string;
  claim: ExtractedClaim;
}

export class ExtractionEngine {
  readonly #database: ExtractionDatabase;
  readonly #models: ExtractionModels;
  readonly #embeddings: ExtractionEmbeddings;

  constructor(
    database: ExtractionDatabase,
    models: ExtractionModels,
    embeddings: ExtractionEmbeddings,
  ) {
    this.#database = database;
    this.#models = models;
    this.#embeddings = embeddings;
  }

  async extract(job: ClaimedJob): Promise<ValidatedExtractionResult> {
    const priorSummaries = await this.#database.priorSummaries(
      job.request.session_id,
      job.segmentId,
    );
    return this.#models.extract(job.request, priorSummaries);
  }

  async resolve(
    job: ClaimedJob,
    extracted: ValidatedExtractionResult,
  ): Promise<PreparedSegment> {
    const contextSpecs: ContextSpec[] = [];
    for (const [index, claim] of extracted.claims.entries()) {
      const supportingClaim = ExtractionEngine.#claimText(claim);
      contextSpecs.push({
        mentionId: `c${index}.subject`,
        role: "subject",
        text: claim.subject,
        supportingClaim,
        claim,
      });
      if (claim.object_entity !== null) {
        contextSpecs.push({
          mentionId: `c${index}.object`,
          role: "object",
          text: claim.object_entity,
          supportingClaim,
          claim,
        });
      }
    }

    const candidateInputs = contextSpecs.map(
      ({ role, text, supportingClaim }) =>
        truncateUtf8Bytes(
          `Entity mention: ${text}\nEndpoint role: ${role}\nSupporting claim: ${supportingClaim}\nSegment summary: ${extracted.summary}`,
          MAX_EMBEDDING_INPUT_BYTES,
        ),
    );
    const mentionEmbeddings = await this.#embeddings.embed(
      candidateInputs,
      "query",
    );
    if (mentionEmbeddings.length !== contextSpecs.length) {
      throw new Error("embedding response did not match entity mentions");
    }
    const candidateSets = await Promise.all(
      contextSpecs.map((spec, index) =>
        this.#database.entityCandidates(spec.text, mentionEmbeddings[index]!),
      ),
    );
    const contexts: MentionContext[] = contextSpecs.map((spec, index) => ({
      mentionId: spec.mentionId,
      role: spec.role,
      text: spec.text,
      supportingClaim: spec.supportingClaim,
      candidates: candidateSets[index]!,
    }));
    const claimsByMentionId = new Map(
      contextSpecs.map((spec) => [spec.mentionId, spec.claim]),
    );

    let plan: ValidatedResolutionPlan = { keptClaims: [], mentions: [] };
    if (contexts.length > 0) {
      plan = await this.#models.resolve(
        job.request,
        extracted.summary,
        extracted.claims,
        contexts,
      );
      console.info("claim triage completed", {
        proposed: extracted.claims.length,
        kept: plan.keptClaims.length,
        dropped: extracted.claims.length - plan.keptClaims.length,
      });
    }

    const entitiesById = new Map<string, PreparedEntity>();
    const occurrenceEntities = new Map<string, PreparedEntity>();
    for (const {
      context,
      resolution,
      selectedCandidate: candidate,
    } of plan.mentions) {
      let entityId: string;
      let canonicalName: string;
      let description: string;
      let aliases: readonly string[];
      let isNew: boolean;
      if (candidate === null) {
        if (resolution.same_new_entity_as === null) {
          canonicalName = context.text;
          description = truncateCodePoints(
            `Entity named ${canonicalName} in claim: ${context.supportingClaim}.`,
            MAX_ENTITY_DESCRIPTION_CODE_POINTS,
          );
          aliases = ExtractionEngine.#uniqueAliases(
            canonicalName,
            context.text,
          );
          entityId = newMentionEntityIdFor(
            job.segmentId,
            job.sourceFingerprint,
            context.mentionId,
            claimsByMentionId.get(context.mentionId)!,
          );
          isNew = true;
        } else {
          const groupedEntity = occurrenceEntities.get(
            resolution.same_new_entity_as,
          );
          if (groupedEntity === undefined) {
            throw new Error(
              `validated new-entity group missing for ${context.mentionId}`,
            );
          }
          entityId = groupedEntity.id;
          canonicalName = groupedEntity.canonicalName;
          description = groupedEntity.description;
          aliases = ExtractionEngine.#uniqueAliases(
            ...groupedEntity.aliases,
            context.text,
          );
          isNew = groupedEntity.isNew;
        }
      } else {
        entityId = candidate.id;
        canonicalName = candidate.canonicalName;
        description = candidate.description;
        aliases = ExtractionEngine.#uniqueAliases(
          canonicalName,
          ...candidate.aliases,
          context.text,
        );
        isNew = false;
      }

      const previous = entitiesById.get(entityId);
      const entity: PreparedEntity =
        previous === undefined
          ? {
              id: entityId,
              canonicalName,
              normalizedName: normalizeName(canonicalName),
              description,
              aliases,
              embedding: null,
              isNew,
            }
          : {
              ...previous,
              aliases: ExtractionEngine.#uniqueAliases(
                ...previous.aliases,
                ...aliases,
              ),
            };
      entitiesById.set(entityId, entity);
      occurrenceEntities.set(context.mentionId, entity);
    }

    let entities = [...entitiesById.values()];
    const newEntities = entities.filter((entity) => entity.isNew);
    const claimTexts = plan.keptClaims.map(({ claim }) =>
      truncateUtf8Bytes(
        ExtractionEngine.#claimText(claim),
        MAX_EMBEDDING_INPUT_BYTES,
      ),
    );
    const entityTexts = newEntities.map((entity) =>
      truncateUtf8Bytes(
        `${entity.canonicalName}: ${entity.description}`,
        MAX_EMBEDDING_INPUT_BYTES,
      ),
    );
    const documentEmbeddings = await this.#embeddings.embed(
      [...claimTexts, ...entityTexts],
      "document",
    );
    if (
      documentEmbeddings.length !==
      plan.keptClaims.length + newEntities.length
    ) {
      throw new Error("embedding response did not match prepared documents");
    }
    const claimEmbeddings = documentEmbeddings.slice(0, plan.keptClaims.length);
    const entityEmbeddings = documentEmbeddings.slice(plan.keptClaims.length);
    const embeddingByEntity = new Map(
      newEntities.map((entity, index) => [entity.id, entityEmbeddings[index]!]),
    );
    entities = entities.map((entity) => ({
      ...entity,
      embedding: embeddingByEntity.get(entity.id) ?? null,
    }));

    const claims: PreparedClaim[] = plan.keptClaims.map(
      ({ index, claim }, embeddingIndex) => {
        const subjectEntity = occurrenceEntities.get(`c${index}.subject`);
        if (subjectEntity === undefined) {
          throw new Error(`missing subject entity for c${index}`);
        }
        const objectEntity =
          claim.object_entity === null
            ? null
            : occurrenceEntities.get(`c${index}.object`);
        if (claim.object_entity !== null && objectEntity === undefined) {
          throw new Error(`missing object entity for c${index}`);
        }
        const objectEntityId = objectEntity?.id ?? null;
        return {
          id: claimIdFor(job.segmentId, index),
          subject: claim.subject,
          subjectEntityId: subjectEntity.id,
          predicate: claim.predicate,
          confidence: claim.confidence,
          objectEntity: claim.object_entity,
          objectEntityId,
          objectValue: claim.object_value,
          equivalenceKey: equivalenceKey(subjectEntity.id, claim.predicate, {
            objectEntityId,
            objectValue: claim.object_value,
          }),
          embedding: claimEmbeddings[embeddingIndex]!,
        };
      },
    );

    return {
      id: job.segmentId,
      sessionId: job.request.session_id,
      startUserMessageId: job.request.start_user_message_id,
      endUserMessageId: job.request.end_user_message_id,
      sourceBoundaryVersion: job.request.source_boundary_version,
      startSourceMessageId: job.request.start_source_message_id,
      endSourceMessageId: job.request.end_source_message_id,
      summary: extracted.summary,
      entities,
      claims,
      projectionVersion: job.request.projection_version,
    };
  }

  static #claimText(claim: ExtractedClaim): string {
    const objectText = claim.object_entity ?? claim.object_value;
    if (objectText === null) {
      throw new Error("extracted claim has no object");
    }
    return `${claim.subject} | ${claim.predicate} | ${objectText}`;
  }

  static #uniqueAliases(...aliases: readonly string[]): string[] {
    const unique = new Map<string, string>();
    for (const alias of aliases) {
      const key = normalizeName(alias);
      if (!unique.has(key)) unique.set(key, alias.trim());
    }
    return [...unique.values()];
  }
}
